internal/flow/promote.go

```diff
@@ -165,33 +158,30 @@ func (p *PromoteEvent) Unmarshal(data []byte) error {
 func (s *Service) ExecPromote(ctx context.Context, p PromoteEvent) error {
 	span, ctx := s.Tracer.FromCtx(ctx, "flow.ExecPromote")
 	defer span.Finish()
-	var result PromoteResult
 	err := s.retry(ctx, func(ctx context.Context, attempt int) (bool, error) {
 		logger := log.WithContext(ctx)
-
-		sourceConfigRepoPath, closeSource, err := git.TempDirAsync(ctx, s.Tracer, "k8s-config-promote-source")
-		if err != nil {
-			return true, err
-		}
-		defer closeSource(ctx)
-
-		// find current released artifact.json for service in env - 1 (dev for staging, staging for prod)
-		logger.Debugf("Cloning source config repo %s into %s", s.Git.ConfigRepoURL, sourceConfigRepoPath)
-		_, err = s.Git.Clone(ctx, sourceConfigRepoPath)
-		if err != nil {
-			return true, errors.WithMessagef(err, "clone into '%s'", sourceConfigRepoPath)
-		}
-
-		hash := plumbing.NewHash(p.Hash)
+		artifactID := p.ArtifactID
 		service := p.Service
 		environment := p.Environment
 		namespace := p.Namespace
 		actor := p.Actor
-		logger.Debugf("internal/flow: Promote: release hash '%v'", hash)
-		err = s.Git.Checkout(ctx, sourceConfigRepoPath, hash)
+
+		var (
+			artifactSourcePath, sourcePath string
+			closeSource                    func(context.Context)
+			err                            error
+		)
+		// when promoting to dev we should look for the artifact instead of
+		// release as the artifact have never been released.
+		if environment == "dev" {
```

Comment:
This is a bit dangerous if we were to make policies that skip environments.
Coming to think of it, isn't it the same as `hamctl release --env dev --branch master`?

Reply:
It is, and is a result of our old Git promote between branches implementation.
It is the same as now, but just moved into this code section.